File: core/src/test/scala/unit/kafka/admin/ListConsumerGroupTest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kafka.admin

import joptsimple.OptionException
import org.junit.jupiter.api.Assertions._
import kafka.utils.{TestInfoUtils, TestUtils}
import org.apache.kafka.clients.admin.ConsumerGroupListing
import org.apache.kafka.common.{ConsumerGroupState, GroupType}
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource

import java.util.Optional

class ListConsumerGroupTest extends ConsumerGroupCommandTest {

  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
  def testListConsumerGroupsWithoutFilters(quorum: String, groupProtocol: String): Unit = {
    val simpleGroup = "simple-group"
    val protocolGroup = "protocol-group"

    createOffsetsTopic()

    addSimpleGroupExecutor(group = simpleGroup)
    addConsumerGroupExecutor(numConsumers = 1)
    addConsumerGroupExecutor(numConsumers = 1, group = protocolGroup, groupProtocol = groupProtocol)

    val cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list")
    val service = getConsumerGroupService(cgcArgs)

    val expectedGroups = Set(protocolGroup, group, simpleGroup)
    var foundGroups = Set.empty[String]
    TestUtils.waitUntilTrue(() => {
      foundGroups = service.listConsumerGroups().toSet
      expectedGroups == foundGroups
    }, s"Expected --list to show groups $expectedGroups, but found $foundGroups.")
  }

  @Test
  def testListWithUnrecognizedNewConsumerOption(): Unit = {
    val cgcArgs = Array("--new-consumer", "--bootstrap-server", bootstrapServers(), "--list")
    assertThrows(classOf[OptionException], () => getConsumerGroupService(cgcArgs))
  }

  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
  def testListConsumerGroupsWithStates(quorum: String, groupProtocol: String): Unit = {
    val simpleGroup = "simple-group"

    createOffsetsTopic()

    addSimpleGroupExecutor(group = simpleGroup)
    addConsumerGroupExecutor(numConsumers = 1)

    val cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--state")
    val service = getConsumerGroupService(cgcArgs)

    var expectedListing = Set(
      new ConsumerGroupListing(
        simpleGroup,
        true,
        Optional.of(ConsumerGroupState.EMPTY),
        Optional.of(GroupType.CLASSIC)
      ),
      new ConsumerGroupListing(
        group,
        false,
        Optional.of(ConsumerGroupState.STABLE),
        Optional.of(GroupType.CLASSIC)
      )
    )

    assertGroupListing(Set.empty, ConsumerGroupState.values.toSet, expectedListing, service)

    expectedListing = Set(
      new ConsumerGroupListing(
        group,
        false,
        Optional.of(ConsumerGroupState.STABLE),
        Optional.of(GroupType.CLASSIC)
      )
    )

    assertGroupListing(Set.empty, Set(ConsumerGroupState.STABLE), expectedListing, service)

    assertGroupListing(Set.empty, Set(ConsumerGroupState.PREPARING_REBALANCE), Set.empty[ConsumerGroupListing], service)
  }

  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersClassicGroupProtocolOnly"))
  def testListConsumerGroupsWithTypesClassicProtocol(quorum: String, groupProtocol: String): Unit = {
    val simpleGroup = "simple-group"

    createOffsetsTopic()

    addSimpleGroupExecutor(group = simpleGroup)
    addConsumerGroupExecutor(numConsumers = 1)

    val cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type")
    val service = getConsumerGroupService(cgcArgs)

    val expectedListing = Set(
      new ConsumerGroupListing(
        simpleGroup,
        true,
        Optional.of(ConsumerGroupState.EMPTY),
        Optional.of(GroupType.CLASSIC)
      ),
      new ConsumerGroupListing(
        group,
        false,
        Optional.of(ConsumerGroupState.STABLE),
        Optional.of(GroupType.CLASSIC)
      )
    )

    // No filters explicitly mentioned. Expectation is that all groups are returned.
    assertGroupListing(Set.empty, Set.empty, expectedListing, service)

    // When group type is mentioned:
    // Old Group Coordinator returns empty listings if the type is not Classic.
    // New Group Coordinator returns groups according to the filter.
    assertGroupListing(Set(GroupType.CONSUMER), Set.empty, Set.empty[ConsumerGroupListing], service)

    assertGroupListing(Set(GroupType.CLASSIC), Set.empty, expectedListing, service)
  }

  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersConsumerGroupProtocolOnly"))
  def testListConsumerGroupsWithTypesConsumerProtocol(quorum: String, groupProtocol: String): Unit = {
    val simpleGroup = "simple-group"
    val protocolGroup = "protocol-group"

    createOffsetsTopic()

    addSimpleGroupExecutor(group = simpleGroup)
    addConsumerGroupExecutor(numConsumers = 1)
    addConsumerGroupExecutor(numConsumers = 1, group = protocolGroup, groupProtocol = groupProtocol)

    val cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type")
    val service = getConsumerGroupService(cgcArgs)

    // No filters explicitly mentioned. Expectation is that all groups are returned.
    var expectedListing = Set(
      new ConsumerGroupListing(
        simpleGroup,
        true,
        Optional.of(ConsumerGroupState.EMPTY),
        Optional.of(GroupType.CLASSIC)
      ),
      new ConsumerGroupListing(
        group,
        false,
        Optional.of(ConsumerGroupState.STABLE),
        Optional.of(GroupType.CLASSIC)
      ),
      new ConsumerGroupListing(
        protocolGroup,
        false,
        Optional.of(ConsumerGroupState.STABLE),
        Optional.of(GroupType.CONSUMER)
      )
    )

    assertGroupListing(Set.empty, Set.empty, expectedListing, service)

    // When group type is mentioned:
    // New Group Coordinator returns groups according to the filter.
    expectedListing = Set(
      new ConsumerGroupListing(
        protocolGroup,
        false,
        Optional.of(ConsumerGroupState.STABLE),
        Optional.of(GroupType.CONSUMER)
      )
    )

    assertGroupListing(Set(GroupType.CONSUMER), Set.empty, expectedListing, service)

    expectedListing = Set(
      new ConsumerGroupListing(
        simpleGroup,
        true,
        Optional.of(ConsumerGroupState.EMPTY),
        Optional.of(GroupType.CLASSIC)
      ),
      new ConsumerGroupListing(
        group,
        false,
        Optional.of(ConsumerGroupState.STABLE),
        Optional.of(GroupType.CLASSIC)
      )
    )

    assertGroupListing(Set(GroupType.CLASSIC), Set.empty, expectedListing, service)
  }

  @Test
  def testConsumerGroupStatesFromString(): Unit = {
    var result = ConsumerGroupCommand.consumerGroupStatesFromString("Stable")
    assertEquals(Set(ConsumerGroupState.STABLE), result)

    result = ConsumerGroupCommand.consumerGroupStatesFromString("Stable, PreparingRebalance")
    assertEquals(Set(ConsumerGroupState.STABLE, ConsumerGroupState.PREPARING_REBALANCE), result)

    result = ConsumerGroupCommand.consumerGroupStatesFromString("Dead,CompletingRebalance,")
    assertEquals(Set(ConsumerGroupState.DEAD, ConsumerGroupState.COMPLETING_REBALANCE), result)

    result = ConsumerGroupCommand.consumerGroupStatesFromString("stable")
    assertEquals(Set(ConsumerGroupState.STABLE), result)

    result = ConsumerGroupCommand.consumerGroupStatesFromString("stable, assigning")
    assertEquals(Set(ConsumerGroupState.STABLE, ConsumerGroupState.ASSIGNING), result)

    result = ConsumerGroupCommand.consumerGroupStatesFromString("dead,reconciling,")
    assertEquals(Set(ConsumerGroupState.DEAD, ConsumerGroupState.RECONCILING), result)

    assertThrows(classOf[IllegalArgumentException], () => ConsumerGroupCommand.consumerGroupStatesFromString("bad, wrong"))

    assertThrows(classOf[IllegalArgumentException], () => ConsumerGroupCommand.consumerGroupStatesFromString("  bad, Stable"))

    assertThrows(classOf[IllegalArgumentException], () => ConsumerGroupCommand.consumerGroupStatesFromString("   ,   ,"))
  }

  @Test
  def testConsumerGroupTypesFromString(): Unit = {
    var result = ConsumerGroupCommand.consumerGroupTypesFromString("consumer")
    assertEquals(Set(GroupType.CONSUMER), result)

    result = ConsumerGroupCommand.consumerGroupTypesFromString("consumer, classic")
    assertEquals(Set(GroupType.CONSUMER, GroupType.CLASSIC), result)

    result = ConsumerGroupCommand.consumerGroupTypesFromString("Consumer, Classic")
    assertEquals(Set(GroupType.CONSUMER, GroupType.CLASSIC), result)

    assertThrows(classOf[IllegalArgumentException], () => ConsumerGroupCommand.consumerGroupTypesFromString("bad, wrong"))

    assertThrows(classOf[IllegalArgumentException], () => ConsumerGroupCommand.consumerGroupTypesFromString("  bad, generic"))

    assertThrows(classOf[IllegalArgumentException], () => ConsumerGroupCommand.consumerGroupTypesFromString("   ,   ,"))
  }

  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersClassicGroupProtocolOnly"))
  def testListGroupCommandClassicProtocol(quorum: String, groupProtocol: String): Unit = {
    val simpleGroup = "simple-group"
    val protocolGroup = "protocol-group"

    createOffsetsTopic()

    addSimpleGroupExecutor(group = simpleGroup)
    addConsumerGroupExecutor(numConsumers = 1)
    addConsumerGroupExecutor(numConsumers = 1, group = protocolGroup, groupProtocol = groupProtocol)
    var out = ""

    var cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      !out.contains("STATE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup and no header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--state")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("STATE") && !out.contains("TYPE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("TYPE") && !out.contains("STATE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--state", "--type")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("TYPE") && out.contains("STATE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--state", "Stable")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("STATE") && out.contains(group) && out.contains("Stable") && out.contains(protocolGroup)
    }, s"Expected to find $group, $protocolGroup in state Stable and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--state", "stable")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("STATE") && out.contains(group) && out.contains("Stable") && out.contains(protocolGroup)
    }, s"Expected to find $group, $protocolGroup in state Stable and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type", "Classic")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("TYPE") && out.contains("Classic") && !out.contains("STATE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type", "classic")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("TYPE") && out.contains("Classic") && !out.contains("STATE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup  and the header, but found $out")
  }

  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersConsumerGroupProtocolOnly"))
  def testListGroupCommandConsumerProtocol(quorum: String, groupProtocol: String): Unit = {
    val simpleGroup = "simple-group"
    val protocolGroup = "protocol-group"

    createOffsetsTopic()

    addSimpleGroupExecutor(group = simpleGroup)
    addConsumerGroupExecutor(numConsumers = 1)
    addConsumerGroupExecutor(numConsumers = 1, group = protocolGroup, groupProtocol = groupProtocol)
    var out = ""

    var cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      !out.contains("STATE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup and no header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--state")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("STATE") && !out.contains("TYPE") && out.contains(simpleGroup) && out.contains(group) && out.contains(protocolGroup)
    }, s"Expected to find $simpleGroup, $group, $protocolGroup and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("TYPE") && out.contains("Consumer") && !out.contains("STATE") && out.contains(protocolGroup) && out.contains(simpleGroup) && out.contains(group)
    }, s"Expected to find $protocolGroup and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type", "consumer")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("TYPE") && out.contains("Consumer") && !out.contains("STATE") && out.contains(protocolGroup)
    }, s"Expected to find $protocolGroup and the header, but found $out")

    cgcArgs = Array("--bootstrap-server", bootstrapServers(), "--list", "--type", "consumer", "--state", "Stable")
    TestUtils.waitUntilTrue(() => {
      out = TestUtils.grabConsoleOutput(ConsumerGroupCommand.main(cgcArgs))
      out.contains("TYPE") && out.contains("Consumer") && out.contains("STATE") && out.contains("Stable") && out.contains(protocolGroup)
    }, s"Expected to find $protocolGroup and the header, but found $out")
  }

  private def assertGroupListing(
    typeFilterSet: Set[GroupType],
    stateFilterSet: Set[ConsumerGroupState],
    expectedListing: Set[ConsumerGroupListing],
    service: ConsumerGroupCommand.ConsumerGroupService
  ): Unit = {
    var foundListing = Set.empty[ConsumerGroupListing]
    TestUtils.waitUntilTrue(() => {
      foundListing = service.listConsumerGroupsWithFilters(typeFilterSet, stateFilterSet).toSet
      expectedListing == foundListing
    }, s"Expected to show groups $expectedListing, but found $foundListing")
  }
}

File: tools/src/test/java/org/apache/kafka/tools/consumergroup/ListConsumerGroupTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.tools.consumergroup;

import joptsimple.OptionException;
import org.apache.kafka.clients.admin.ConsumerGroupListing;
import org.apache.kafka.common.ConsumerGroupState;
import org.apache.kafka.test.TestUtils;
import org.apache.kafka.tools.consumergroup.ConsumerGroupCommand.ConsumerGroupService;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class ListConsumerGroupTest extends ConsumerGroupCommandTest {
    @ParameterizedTest
    @ValueSource(strings = {"zk", "kraft"})
    public void testListConsumerGroups(String quorum) throws Exception {
        String simpleGroup = "simple-group";
        addSimpleGroupExecutor(simpleGroup);
        addConsumerGroupExecutor(1);

        String[] cgcArgs = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list"};
        ConsumerGroupService service = getConsumerGroupService(cgcArgs);
        Set<String> expectedGroups = new HashSet<>(Arrays.asList(GROUP, simpleGroup));
        final Set[] foundGroups = new Set[]{Collections.emptySet()};
        TestUtils.waitForCondition(() -> {
            foundGroups[0] = new HashSet<>(service.listConsumerGroups());
            return Objects.equals(expectedGroups, foundGroups[0]);
        }, "Expected --list to show groups " + expectedGroups + ", but found " + foundGroups[0] + ".");
    }

    @ParameterizedTest
    @ValueSource(strings = {"zk", "kraft"})
    public void testListWithUnrecognizedNewConsumerOption() {
        String[] cgcArgs = new String[]{"--new-consumer", "--bootstrap-server", bootstrapServers(listenerName()), "--list"};
        assertThrows(OptionException.class, () -> getConsumerGroupService(cgcArgs));
    }

    @ParameterizedTest
    @ValueSource(strings = {"zk", "kraft"})
    public void testListConsumerGroupsWithStates() throws Exception {
        String simpleGroup = "simple-group";
        addSimpleGroupExecutor(simpleGroup);
        addConsumerGroupExecutor(1);

        String[] cgcArgs = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--state"};
        ConsumerGroupService service = getConsumerGroupService(cgcArgs);

        Set<ConsumerGroupListing> expectedListing = new HashSet<>(Arrays.asList(
            new ConsumerGroupListing(simpleGroup, true, Optional.of(ConsumerGroupState.EMPTY)),
            new ConsumerGroupListing(GROUP, false, Optional.of(ConsumerGroupState.STABLE))));

        final Set[] foundListing = new Set[]{Collections.emptySet()};
        TestUtils.waitForCondition(() -> {
            foundListing[0] = new HashSet<>(service.listConsumerGroupsWithState(new HashSet<>(Arrays.asList(ConsumerGroupState.values()))));
            return Objects.equals(expectedListing, foundListing[0]);
        }, "Expected to show groups " + expectedListing + ", but found " + foundListing[0]);

        Set<ConsumerGroupListing> expectedListingStable = Collections.singleton(
            new ConsumerGroupListing(GROUP, false, Optional.of(ConsumerGroupState.STABLE)));

        foundListing[0] = Collections.emptySet();

        TestUtils.waitForCondition(() -> {
            foundListing[0] = new HashSet<>(service.listConsumerGroupsWithState(Collections.singleton(ConsumerGroupState.STABLE)));
            return Objects.equals(expectedListingStable, foundListing[0]);
        }, "Expected to show groups " + expectedListingStable + ", but found " + foundListing[0]);
    }

    @ParameterizedTest
    @ValueSource(strings = {"zk", "kraft"})
    public void testConsumerGroupStatesFromString(String quorum) {
        Set<ConsumerGroupState> result = ConsumerGroupCommand.consumerGroupStatesFromString("Stable");
        assertEquals(Collections.singleton(ConsumerGroupState.STABLE), result);

        result = ConsumerGroupCommand.consumerGroupStatesFromString("Stable, PreparingRebalance");
        assertEquals(new HashSet<>(Arrays.asList(ConsumerGroupState.STABLE, ConsumerGroupState.PREPARING_REBALANCE)), result);

        result = ConsumerGroupCommand.consumerGroupStatesFromString("Dead,CompletingRebalance,");
        assertEquals(new HashSet<>(Arrays.asList(ConsumerGroupState.DEAD, ConsumerGroupState.COMPLETING_REBALANCE)), result);

        assertThrows(IllegalArgumentException.class, () -> ConsumerGroupCommand.consumerGroupStatesFromString("bad, wrong"));

        assertThrows(IllegalArgumentException.class, () -> ConsumerGroupCommand.consumerGroupStatesFromString("stable"));

        assertThrows(IllegalArgumentException.class, () -> ConsumerGroupCommand.consumerGroupStatesFromString("  bad, Stable"));

        assertThrows(IllegalArgumentException.class, () -> ConsumerGroupCommand.consumerGroupStatesFromString("   ,   ,"));
    }

    @ParameterizedTest
    @ValueSource(strings = {"zk", "kraft"})
    public void testListGroupCommand(String quorum) throws Exception {
        String simpleGroup = "simple-group";
        addSimpleGroupExecutor(simpleGroup);
        addConsumerGroupExecutor(1);
        final String[] out = {""};

        String[] cgcArgs1 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list"};
        TestUtils.waitForCondition(() -> {
            out[0] = TestUtils.grabConsoleOutput(() -> ConsumerGroupCommand.main(cgcArgs1));
            return !out[0].contains("STATE") && out[0].contains(simpleGroup) && out[0].contains(GROUP);
        }, "Expected to find " + simpleGroup + ", " + GROUP + " and no header, but found " + out[0]);

        String[] cgcArgs2 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--state"};
        TestUtils.waitForCondition(() -> {
            out[0] = TestUtils.grabConsoleOutput(() -> ConsumerGroupCommand.main(cgcArgs2));
            return out[0].contains("STATE") && out[0].contains(simpleGroup) && out[0].contains(GROUP);
        }, "Expected to find " + simpleGroup + ", " + GROUP + " and the header, but found " + out[0]);

        String[] cgcArgs3 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--state", "Stable"};
        TestUtils.waitForCondition(() -> {
            out[0] = TestUtils.grabConsoleOutput(() -> ConsumerGroupCommand.main(cgcArgs3));
            return out[0].contains("STATE") && out[0].contains(GROUP) && out[0].contains("Stable");
        }, "Expected to find " + GROUP + " in state Stable and the header, but found " + out[0]);
    }
}
